Connector Generated ID is not always consistent

**Problem description**
The Flink Connector generates an ID which is used as the MasterRestoreHook.  On recovery this number is used to load the stored state from the Savepoint into the Flink Connector.

Mostly this works, however on rare occasions the Flink Connector appears to generate a different ID and therefore the MasterRestoreHook fails to recover.

```
java.lang.RuntimeException: org.apache.flink.runtime.client.JobExecutionException: Could not set up JobManager
 at org.apache.flink.util.function.CheckedSupplier.lambda$unchecked$0(CheckedSupplier.java:36)
 at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1590)
 at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:39)
 at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:415)
 at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
 at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
 at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
 at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Caused by: org.apache.flink.runtime.client.JobExecutionException: Could not set up JobManager
 at org.apache.flink.runtime.jobmaster.JobManagerRunner.<init>(JobManagerRunner.java:176)
 at org.apache.flink.runtime.dispatcher.Dispatcher$DefaultJobManagerRunnerFactory.createJobManagerRunner(Dispatcher.java:1058)
 at org.apache.flink.runtime.dispatcher.Dispatcher.lambda$createJobManagerRunner$5(Dispatcher.java:308)
 at org.apache.flink.util.function.CheckedSupplier.lambda$unchecked$0(CheckedSupplier.java:34)
 ... 7 more
Caused by: java.lang.IllegalStateException: Found state '-1347114770' which is not resumed by any hook.
 at org.apache.flink.runtime.checkpoint.hooks.MasterHooks.restoreMasterHooks(MasterHooks.java:258)
 at org.apache.flink.runtime.checkpoint.CheckpointCoordinator.restoreLatestCheckpointedState(CheckpointCoordinator.java:1053)
 at org.apache.flink.runtime.checkpoint.CheckpointCoordinator.restoreSavepoint(CheckpointCoordinator.java:1114)
 at org.apache.flink.runtime.jobmaster.JobMaster.tryRestoreExecutionGraphFromSavepoint(JobMaster.java:1241)
 at org.apache.flink.runtime.jobmaster.JobMaster.createAndRestoreExecutionGraph(JobMaster.java:1165)
 at org.apache.flink.runtime.jobmaster.JobMaster.<init>(JobMaster.java:296)
 at org.apache.flink.runtime.jobmaster.JobManagerRunner.<init>(JobManagerRunner.java:157)
 ... 10 more
```

**Problem location**
It looks like the issue might be with the generation AbstractStreamingReaderBuilder
```
/**
 * Generate a UID for the source, to distinguish the state associated with the checkpoint hook.  A good generated UID will:
 * 1. be stable across savepoints for the same inputs
 * 2. disambiguate one source from another (e.g. in a program that uses numerous instances of {@link FlinkPravegaReader})
 * 3. allow for reconfiguration of the timeouts
 */
String generateUid() {
    StringBuilder sb = new StringBuilder();
    sb.append(readerGroupScope).append('\n');
    resolveStreams().forEach(s -> sb
            .append(s.getStream().getScopedName())
            .append('/').append(s.getFrom().hashCode())
            .append('/').append(s.getTo().hashCode())
            .append('\n'));
    return Integer.toString(sb.toString().hashCode());
}
```
**Suggestions for an improvement**
The generated ID should always be exactly the same every time the connector is invoked.

--- a/src/main/java/io/pravega/connectors/flink/AbstractStreamingReaderBuilder.java
+++ b/src/main/java/io/pravega/connectors/flink/AbstractStreamingReaderBuilder.java
@@ -190,15 +190,13 @@ abstract class AbstractStreamingReaderBuilder<T, B extends AbstractStreamingRead
      * Generate a UID for the source, to distinguish the state associated with the checkpoint hook.  A good generated UID will:
      * 1. be stable across savepoints for the same inputs
      * 2. disambiguate one source from another (e.g. in a program that uses numerous instances of {@link FlinkPravegaReader})
-     * 3. allow for reconfiguration of the timeouts
+     * 3. allow for reconfiguration of the stream cuts and timeouts
      */
     String generateUid() {
         StringBuilder sb = new StringBuilder();
         sb.append(readerGroupScope).append('\n');
         resolveStreams().forEach(s -> sb
                 .append(s.getStream().getScopedName())
-                .append('/').append(s.getFrom().hashCode())
-                .append('/').append(s.getTo().hashCode())
                 .append('\n'));
         return Integer.toString(sb.toString().hashCode());
     }

--- a/src/test/java/io/pravega/connectors/flink/FlinkPravegaReaderTest.java
+++ b/src/test/java/io/pravega/connectors/flink/FlinkPravegaReaderTest.java
@@ -80,8 +80,10 @@ public class FlinkPravegaReaderTest {
 
     private static final String SAMPLE_SCOPE = "scope";
     private static final String SAMPLE_STREAM_NAME = "stream";
+    private static final String SAMPLE_STREAM_NAME_2 = "stream-2";
     private static final String SAMPLE_COMPLETE_STREAM_NAME = SAMPLE_SCOPE + '/' + SAMPLE_STREAM_NAME;
     private static final Stream SAMPLE_STREAM = Stream.of(SAMPLE_SCOPE, SAMPLE_STREAM_NAME);
+    private static final Stream SAMPLE_STREAM_2 = Stream.of(SAMPLE_SCOPE, SAMPLE_STREAM_NAME_2);
     private static final Segment SAMPLE_SEGMENT = new Segment(SAMPLE_SCOPE, SAMPLE_STREAM.getStreamName(), 1);
     private static final StreamCut SAMPLE_CUT = new StreamCutImpl(SAMPLE_STREAM, Collections.singletonMap(SAMPLE_SEGMENT, 42L));
     private static final StreamCut SAMPLE_CUT2 = new StreamCutImpl(SAMPLE_STREAM, Collections.singletonMap(SAMPLE_SEGMENT, 1024L));
@@ -413,23 +415,17 @@ public class FlinkPravegaReaderTest {
 
         TestableStreamingReaderBuilder builder2 = new TestableStreamingReaderBuilder()
                 .withReaderGroupScope(SAMPLE_SCOPE)
-                .forStream(SAMPLE_STREAM, SAMPLE_CUT, StreamCut.UNBOUNDED)
+                .forStream(SAMPLE_STREAM, SAMPLE_CUT, SAMPLE_CUT2)
                 .withEventReadTimeout(Time.seconds(42L));
         String uid2 = builder2.generateUid();
 
         TestableStreamingReaderBuilder builder3 = new TestableStreamingReaderBuilder()
                 .withReaderGroupScope(SAMPLE_SCOPE)
-                .forStream(SAMPLE_STREAM, SAMPLE_CUT2, StreamCut.UNBOUNDED);
+                .forStream(SAMPLE_STREAM_2);
         String uid3 = builder3.generateUid();
-
-        TestableStreamingReaderBuilder builder4 = new TestableStreamingReaderBuilder()
-                .withReaderGroupScope(SAMPLE_SCOPE)
-                .forStream(SAMPLE_STREAM, SAMPLE_CUT, SAMPLE_CUT2);
-        String uid4 = builder4.generateUid();
 
         assertEquals(uid1, uid2);
         assertNotEquals(uid1, uid3);
-        assertNotEquals(uid1, uid4);
     }
 
     // endregion

